### PR TITLE
Peer shrinkwrap fix

### LINKED ISSUE #5135: peerDependencies and npm shrinkwrap
If I have an app[1] that requires a module, in this case `express@~3.4.8` and it also has a module[2] that requires that same version. That module[2] also requires another module[3] that requires modules (`express-state` & `express-map`) with a peerDependency on `express@3.x`.

If I install this, as is, then it will work perfectly:
https://gist.github.com/davglass/11225979#file-no-shrinkwrap-txt

However, if I add a shrinkwrap file:
https://github.com/davglass/peer-app/blob/master/npm-shrinkwrap.json

Then try to install it:
https://gist.github.com/davglass/11225979#file-shrinkwrap-txt

It will not recognize that `express@3.4.8` was installed at a higher level and it will pull in the latest `express@3.5.1` which will break the application. 

What I think is happening here is that when processing the shrinkwrap file, the peerDependencies logic kicks in and says "hey, I need `express@3.x` and it's not at this level, go fetch it" when it should check to see if it's available in the tree above it. This is what "I think" `npm install` does for us. But since that logic is skipped for the shrinkwrap, it's actually doing the wrong thing.

Links:
[1]app: https://github.com/davglass/peer-app
[2]module1: https://github.com/davglass/module1
[3]module2: https://github.com/davglass/module1

Possibly related to #5080 

I'm happy to dive in and see if I can find the root cause of this issue but I wanted to confirm first that this is an issue.

If #5080 removes the installation of peerDependencies and just warns, then I think this issue will go away since the install won't happen.

### PR DESCRIPTION
**Pull request by ghost**
This should fix #5135.  

I read the guidelines for contributing which recommended squashing to a single commit.  I intentionally left two commits so that you can verify the added test is failing first.  I was just bitten by not doing this, as I was comparing two undefined values output by the rmFrom function.  There was also another test that was fixed, that was also comparing two undefined values.

In response to [CERT vulnerability note VU#319816](https://www.kb.cert.org/vuls/id/319816) we started requiring the use of npm-shrinkwrap.json files in our builds. As this issue circumvents this precaution while also breaking applications, we are anxious to see this issue resolved in the next lts release.

### CODE REVIEW

**Comment by majgis** [comment]
The following provides some insight into what the problem was:

```bash
$ grep -r -A 3 "\"peerDependencies\"" node_modules/
node_modules//react-bootstrap/node_modules/react-dom/package.json:  "peerDependencies": {
node_modules//react-bootstrap/node_modules/react-dom/package.json-    "react": "^15.0.1"
node_modules//react-bootstrap/node_modules/react-dom/package.json-  },
node_modules//react-bootstrap/node_modules/react-dom/package.json-  "scripts": {
--
node_modules//react-bootstrap/node_modules/react-overlays/package.json:  "peerDependencies": {
node_modules//react-bootstrap/node_modules/react-overlays/package.json-    "react": ">=0.14.0",
node_modules//react-bootstrap/node_modules/react-overlays/package.json-    "react-dom": ">=0.14.0"
node_modules//react-bootstrap/node_modules/react-overlays/package.json-  },
--
node_modules//react-bootstrap/node_modules/uncontrollable/package.json:  "peerDependencies": {
node_modules//react-bootstrap/node_modules/uncontrollable/package.json-    "react": ">=0.11.0"
node_modules//react-bootstrap/node_modules/uncontrollable/package.json-  },
node_modules//react-bootstrap/node_modules/uncontrollable/package.json-  "release-script": {
--
node_modules//react-bootstrap/package.json:  "peerDependencies": {
node_modules//react-bootstrap/package.json-    "react": ">=0.14.0",
node_modules//react-bootstrap/package.json-    "react-dom": ">=0.14.0"
node_modules//react-bootstrap/package.json-  },
--
node_modules//react-dom/package.json:  "peerDependencies": {
node_modules//react-dom/package.json-    "react": "^0.14.8"
node_modules//react-dom/package.json-  },
node_modules//react-dom/package.json-  "scripts": {
```

File: test/common.js [hunk 1]
```diff
@@ -1,7 +1,10 @@
-
```
Comment (iarna) [inline test/common.js]: This file isn't, as far as I can tell, used at all any more. It should really go away, I wouldn't extend it. Just put your function directly into `common-tap` above. All the tests should really be loading `common-tap` anyway, as it does some important environment configuration.

File: test/packages/npm-test-shrinkwrap/test.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-var path = require("path")
```
Comment (iarna) [inline test/packages/npm-test-shrinkwrap/test.js]: This file is a part of the legacy tests and... as best as I can tell shouldn't actually be in the branch at all? Regardless, the one that's used now is [test/tap/legacy-shrinkwrap.js](https://github.com/npm/npm/blob/lts/test/tap/legacy-shrinkwrap.js)

File: test/common.js [hunk 3]
```diff
@@ -1,7 +1,10 @@
...
-// whatever, it's just tests.
-;["util","assert"].forEach(function (thing) {
-  thing = require("thing")
-  for (var i in thing) global[i] = thing[i]
-}
-
+exports.rmFrom = function rmFrom (obj) {
```
Comment (iarna) [inline test/common.js]: While I think `rmFrom` is fine when the function is declared near its context (as it was previously), as a globally available tool I'd like the function name to be more descriptive. `rmFromInShrinkwrap` or something like that?

Comment (majgis) [inline test/common.js]: I will change the name to rmFromInShrinkwrap.

File: test/common.js (hunk 1 above)
Comment (majgis) [inline test/common.js]: I will remove test/common.js

File: test/packages/npm-test-shrinkwrap/test.js (hunk 2 above)
Comment (majgis) [inline test/packages/npm-test-shrinkwrap/test.js]: test/tap/legacy-shrinkwrap.js below had it's behavior changed

I will delete this file after confirming it is dead code.

**Comment by ericdfields** [comment]
Just caught this too in our CI environment. Thanks for addressing it.

**Comment by joelzimmer** [comment]
As someone who is affected by it, any ETA on when this would be merged into npm2?

**Comment by majgis** [comment]
@joelzimmer If you or anyone else could confirm that this change works, without causing adverse effects, it might help it along:

To install npm with this fix:
```
npm i -g git+https://github.com/majgis/npm.git#366c0728a1a8865d265d5a9f06389fe0b5dd8183
```

To restore to the latest official version:
```
npm i -g npm@2
```

**Comment by joelzimmer** [comment]
I was able to confirm that this solves the issue we were experiencing (referenced in #5135.)

**Comment by robrichard** [comment]
I can confirm that this also solves https://github.com/npm/npm/issues/12276

**Comment by kevinold** [comment]
@majgis I can confirm that this solves #12276

**Comment by jvivs** [comment]
Any guidance on when this would make it into an npm LTS release, and then the subsequent node LTS release?

**Comment by majgis** [comment]
@iarna Thank you for your earlier review.  Is there anything else I can do on this PR?

**Comment by iarna** [comment]
LGTM. As for timing, I would expect it to go out in the next `lts` release– maybe next week? @zkat is the owner of that.

**Comment by baens** [comment]
@zkat & @iarna: looks like there was just a release of npm & node but this didn't make the cut. Any idea if this can be in the next minor bump?

**Comment by iarna** [comment]
@baens npm has releases every week, FYI. Looks like I'm responsible for both releases this week, so I'm scheduling this for the lts release. =)

**Comment by baens** [comment]
Awesome @iarna, thank you!

**Comment by iarna** [comment]
And this landed in 2.15.6! =)

**Comment by joelzimmer** [comment]
@iarna thanks so much! When will this be tagged as part of `lts`?

**Comment by baens** [comment]
HI @iarna, any ETA on when `2.15.6` will land on the node `v4.x-staging` branch with this fix in it?

**Pull request closed without merging**